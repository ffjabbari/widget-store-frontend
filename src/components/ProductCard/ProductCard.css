/* .ProductCard {
  min-height: 375px;
} */

.ProductCard {
  min-width: 318px;
  min-height: 350px;
  max-width: 318px;
  max-height: 350px;
  background: whitesmoke;
  overflow: hidden;
}

.ProductCard .img-wrapper {
  min-height: 200px;
  max-height: 200px;
  overflow: hidden;
}
