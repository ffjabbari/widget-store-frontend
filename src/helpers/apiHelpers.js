/** API Helper functions. */

// local imports
import { asyncAPIRequest } from '../hooks/apiHook';

/**
 * updateUserProfile()
 *
 * Make patch request to update a user profile. Used by CustomerAddressForm
 * to allow updating customer address of signed in users.
 *
 * @param {object} values - form values
 * @param {object} user - user object
 * @param {token} token - user token
 * @param {function} setResponseError - state setter for error display
 * @param {any} defaultAddress - Falsey or obj with address components
 * @param {function} setFormDisabled - state setter to hide form after update
 *
 * returns: true if an error happened else undefined
 */

export const updateUserProfile = async (
  values,
  user,
  token,
  setResponseError,
  defaultAddress,
  setFormDisabled
) => {
  const userUpdateData = { ...values };
  // remove email to dis-allow email updating from here
  delete userUpdateData['email'];
  delete userUpdateData['user_id'];

  // remomve key/value pairs having a value of empty string from userUpdateData
  Object.entries(userUpdateData).forEach(([key, value]) => {
    if (value === '') delete userUpdateData[key];
  });

  try {
    // make patch request
    const resp = await asyncAPIRequest(
      `/users/${user.username}`,
      'patch',
      userUpdateData,
      { _token: token }
    );
    if (resp.error) {
      setResponseError(resp.error.response.data.message);
      return true;
    } else {
      defaultAddress.current = values;
      setFormDisabled(true);
    }
  } catch (error) {
    setResponseError(error.message);
    return true;
  }
};
