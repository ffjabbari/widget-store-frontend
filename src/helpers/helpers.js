import { Redirect } from 'react-router-dom';
import { motion } from 'framer-motion';

/**
 * Given a path return the "root" argument.
 * For example the "root" of /shop/product/206 would be "/shop".
 *
 * @param {string} path
 */

export const getPathRoot = path => {
  return path.match(/^(\/\w*)\/?.*$/)[1];
};

export const FramerRedirect = props => (
  <motion.div exit="undefined">
    <Redirect {...props} />
  </motion.div>
);

export const animateVariant = (setterFunct, delay, callback, args = []) => {
  // if animation state is false set to 'default' else 'active'
  setterFunct(state => (state === false ? 'default' : 'active'));
  setTimeout(() => {
    setterFunct('default');
    if (callback) callback(...args);
  }, delay);
};
