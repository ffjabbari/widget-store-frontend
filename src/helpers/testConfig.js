/** Test data setup/teardown helper hooks */

export const TEST_DATA = {};

export const populateTestDataHook = testData => {
  testData.product = {
    byline: 'Latest generation smart televison 2021',
    date_added: '2020-12-29T08:00:00.000Z',
    description:
      'Lorem ipsum dolor sit, amet consectetur adipisicing elit. Nihil, voluptates enim nam voluptatum, quia doloribus quis sequi, porro consequatur excepturi eos error voluptas nobis modi atque magni commodi amet placeat.',
    discount: '0.00',
    image_url:
      'https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg',
    name: 'Samsung XG-900 55 inch LCD TV',
    price: '400.40',
    product_id: 1,
    quantity: 100,
    rating: null,
  };

  testData.product2 = {
    byline: 'A good remote.',
    discount: '0.15',
    image_url:
      'https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg',
    name: 'Logitech Singularity Universal Remote',
    price: '40.20',
    product_id: 2,
    rating: null,
    quantity: 1,
  };

  testData.userProfileData = {
    data: {
      user: {
        user_id: 1,
        username: 'widgetLord',
        email: 'foo@gmail.com',
        first_name: 'bob',
        last_name: 'bobert',
        address: '123 Hill Ave.',
        address_line2: null,
        city: 'Big City',
        state: 'Texas',
        postal_code: '33333',
        phone_number: null,
        avatar_url: null,
        orders: [
          {
            order_id: 1,
            order_date: '2021-01-24T08:00:00.000Z',
            status: 'Pending',
          },
        ],
      },
    },
  };

  testData.CustomerInfoFormsValues = {
    first_name: 'Test',
    last_name: 'User',
    email: 'foo@gmail.com',
    address: '123 Main St',
    address_line2: '',
    city: 'Big City',
    state: 'TX',
    postal_code: '39488',
    phone_number: '(555) 555-5555',
    set_default: false,
  };

  testData.cart = {
    items: {
      2: {
        byline: 'The only remote you will need',
        date_added: '2021-01-24T08:00:00.000Z',
        description: 'Lorem ipsum dolor sit, placeat.',
        discount: '0.15',
        image_url:
          'https://cnet3.cbsistatic.com/img/SB-Lc9lni02k-GQeFZlVGII6cpg=/1070x602/2010/08/03/57369690-bb77-11e2-8a8e-0291187978f3/33981226-2-1333-OVR-1.jpg',
        name: 'Logitech Singularity Universal Remote',
        price: '40.20',
        product_id: 2,
        quantity: 2,
        rating: null,
      },
    },
    numCartItems: 1,
    subtotal: '34.17',
  };
};
