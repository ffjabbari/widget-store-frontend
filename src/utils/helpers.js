import { Redirect, Link } from 'react-router-dom';
import { motion } from 'framer-motion';

/**
 * getPathRoot()
 *
 * Given a path return the "root" argument.
 * For example the "root" of /shop/product/206 would be "/shop".
 *
 * @param {string} path
 */

export const getPathRoot = path => {
  return path.match(/^(\/\w*)\/?.*$/)[1];
};

/**
 * <FramerRedirect to="" />
 *
 * Component to force react-router-dom redirect when framer-motion
 * prevents a redirect from executing.
 *
 * If regular <Redirect /> fails in this environment use this redirect instead.
 *
 * @param {any} props
 *
 * returns: motion.div wrapped redirect component
 */
export const FramerRedirect = props => (
  <motion.div exit="undefined">
    <Redirect {...props} />
  </motion.div>
);
export const FramerLink = props => (
  <motion.div exit="undefined">
    <Link {...props} />
  </motion.div>
);

/**
 * animateVariant()
 *
 * Animation function to turn on an animation state then turn it back
 * off after a delay. The state will be set "active" then back to "default".
 *
 * If inital state provided by setterFunct is false this function will skip
 * setting the animation state "active" instead set it to "default".
 *
 * If a callback is provided it will be called with any arguments
 * after setting state back to "default".
 *
 *
 * @param {function} setterFunct - useState setter funct to set animation state
 * @param {integer} delay - delay of setterFunct being called
 * @param {funct} callback - function to call upon completion
 * @param {array} args - array of arguments for callback function
 */

export const animateVariant = (setterFunct, delay, callback, args = []) => {
  // if animation state is false set to 'default' else 'active'
  setterFunct(state => (state === false ? 'default' : 'active'));
  // after delay set back to "default"
  return setTimeout(() => {
    setterFunct('default');
    if (callback) callback(...args);
  }, delay);
};
