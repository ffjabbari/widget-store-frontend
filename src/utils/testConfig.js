/** Test data setup/teardown helper hooks */

export const TEST_DATA = {};

export const populateTestDataHook = testData => {
  testData.product = {
    byline: 'Latest generation smart televison 2021',
    date_added: '2020-12-29T08:00:00.000Z',
    description:
      'Lorem ipsum dolor sit, amet consectetur adipisicing elit. Nihil, voluptates enim nam voluptatum, quia doloribus quis sequi, porro consequatur excepturi eos error voluptas nobis modi atque magni commodi amet placeat.',
    discount: '0.00',
    image_url:
      'https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg',
    name: 'Samsung XG-900 55 inch LCD TV',
    price: '400.40',
    product_id: 1,
    quantity: 100,
    rating: null,
  };

  testData.product2 = {
    byline: 'The only remote you will need',
    description: 'Lorem ipsum dolor sit, placeat.',
    discount: '0.15',
    image_url:
      'https://cnet3.cbsistatic.com/img/SB-Lc9lni02k-GQeFZlVGII6cpg=/1070x602/2010/08/03/57369690-bb77-11e2-8a8e-0291187978f3/33981226-2-1333-OVR-1.jpg',
    name: 'Logitech Singularity Universal Remote',
    price: '40.20',
    product_id: 2,
    quantity: 1,
    rating: null,
  };

  testData.userProfileData = {
    data: {
      user: {
        user_id: 1,
        username: 'widgetLord',
        email: 'foo@gmail.com',
        first_name: 'bob',
        last_name: 'bobert',
        address: '123 Hill Ave.',
        address_line2: null,
        city: 'Big City',
        state: 'Texas',
        postal_code: '33333',
        phone_number: null,
        avatar_url: null,
        orders: [
          {
            order_id: 1,
            order_date: '2021-01-24T08:00:00.000Z',
            status: 'Pending',
          },
        ],
      },
    },
  };

  testData.CustomerInfoFormsValues = {
    first_name: 'Test',
    last_name: 'User',
    email: 'foo@gmail.com',
    address: '123 Main St',
    address_line2: 'Apt 5',
    city: 'Big City',
    state: 'TX',
    postal_code: '39488',
    phone_number: '(555) 555-5555',
  };

  testData.cart = {
    items: {
      2: {
        byline: 'The only remote you will need',
        description: 'Lorem ipsum dolor sit, placeat.',
        discount: '0.15',
        image_url:
          'https://cnet3.cbsistatic.com/img/SB-Lc9lni02k-GQeFZlVGII6cpg=/1070x602/2010/08/03/57369690-bb77-11e2-8a8e-0291187978f3/33981226-2-1333-OVR-1.jpg',
        name: 'Logitech Singularity Universal Remote',
        price: '40.20',
        product_id: 2,
        quantity: 1,
        rating: null,
      },
    },
    numCartItems: 1,
    subtotal: '34.17',
  };

  // mock return from /order/:id
  testData.orderGet = {
    customer: 1,
    customer_info: {
      first_name: 'Test',
      last_name: 'User',
      email: 'foo@gmail.com',
      address: '123 Main St',
      address_line2: '',
      city: 'Big City',
      state: 'TX',
      postal_code: '39488',
      phone_number: '(555) 555-5555',
    },

    discount: null,
    items: [
      { 0: { order_id: 2, product_id: 6, quantity: 1 } },
      { 1: { order_id: 2, product_id: 8, quantity: 1 } },
    ],

    order_date: '2021-02-05T08:00:00.000Z',
    order_id: 2,
    shipping_address: null,
    shipping_cost: '12.00',
    shipping_method: {
      details: { cost: '12.00', name: 'USPS ground shipping (3-7 days)' },
      shipping_method: 'usps_ground',
    },

    status: 'Pending',
    subtotal: '784.16',
    tax: '66.65',
    total: '862.81',
    total_items_quantity: 2,
    tracking_number: null,
  };

  TEST_DATA.shippingMethods = {
    usps_ground: {
      cost: '12.00',
      name: 'USPS ground shipping (3-7 days)',
    },
    usps_priority: {
      cost: '16.00',
      name: 'USPS priority shipping (2-4 days)',
    },
    usps_next_day: {
      cost: '25.00',
      name: 'USPS next day delivery (1 day)',
    },
    ups_ground: {
      cost: '16.00',
      name: 'UPS ground shipping (3-6 days)',
    },
    ups_2nd_day: {
      cost: '20.00',
      name: 'UPS second day delivery (2 days)',
    },
    ups_overnight: {
      cost: '28.00',
      name: 'UPS overnight delivery (1 day)',
    },
  };
};
