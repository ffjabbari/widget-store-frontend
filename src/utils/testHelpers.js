import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { testStore } from '../redux/store/reduxStore';

export const renderWithStore = component =>
  render(
    <Provider store={testStore}>
      <MemoryRouter>{component}</MemoryRouter>
    </Provider>
  );
