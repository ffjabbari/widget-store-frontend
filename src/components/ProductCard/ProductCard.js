import React from 'react';
import { useLocation } from 'react-router-dom';
import Card from 'react-bootstrap/Card';

//local imports
import Pricing1 from '../common/Pricing/Pricing1';
import { getPathRoot } from '../../utils/helpers';
import { FramerLink } from '../../utils/helpers';
import './ProductCard.css';

const ProductCard = ({ product }) => {
  // const history = useHistory();
  const locaton = useLocation();

  // const gotoProductDetail = id => {
  //   history.push(`${getPathRoot(locaton.pathname)}/product/${id}`);
  // };

  // include product byline if product name is not too long (prevent card overflow)
  const productByline = product.name.length < 31 ? product.byline : '';

  return (
    <FramerLink
      to={`${getPathRoot(locaton.pathname)}/product/${product.product_id}`}
      wrapper_className="mx-auto d-inline-block"
      className="darkLink"
    >
      <Card className="ProductCard mb-3">
        <div className="img-wrapper">
          <Card.Img src={product.image_url} alt={product.name} />
        </div>

        <Card.Body>
          <Card.Title tag={product.name.length > 60 ? 'h6' : 'h5'}>
            {product.name}
          </Card.Title>
          <Card.Subtitle tag="h6" className="mb-2 text-muted">
            <em>{productByline}</em>
          </Card.Subtitle>
          <Card.Text>
            <Pricing1 product={product} />
          </Card.Text>
        </Card.Body>
      </Card>
    </FramerLink>
  );
};

export default React.memo(ProductCard);
