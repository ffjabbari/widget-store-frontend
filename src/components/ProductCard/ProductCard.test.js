/** ProductCard tests */

// local imports
import { renderWithStore } from '../../helpers/testHelpers';
import { TEST_DATA, populateTestDataHook } from '../../helpers/testConfig';
import ProductCard from './ProductCard';

beforeAll(() => {
  populateTestDataHook(TEST_DATA);
});

test('renders ProductCard', async () => {
  renderWithStore(<ProductCard product={TEST_DATA.product} />);
});

test('ProductCard snapshot', async () => {
  const { asFragment } = renderWithStore(
    <ProductCard product={TEST_DATA.product} />
  );
  expect(asFragment()).toMatchSnapshot();
});

test('Name, price, and image element in document', async () => {
  const { getByText, getByAltText } = renderWithStore(
    <ProductCard product={TEST_DATA.product} />
  );

  const productName = getByText('Samsung XG-900 55 inch LCD TV');
  expect(productName).toBeInTheDocument();

  const price = getByText(/400.40/);
  expect(price).toBeInTheDocument();

  const img = getByAltText(TEST_DATA.product.name);
  expect(img).toBeInTheDocument();
});
