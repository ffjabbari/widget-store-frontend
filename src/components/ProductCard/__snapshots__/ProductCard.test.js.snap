// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard snapshot 1`] = `
<DocumentFragment>
  <div
    class="ProductCard mb-4 mx-auto card"
  >
    <div
      class="img-wrapper"
    >
      <img
        alt="Samsung XG-900 55 inch LCD TV"
        class="card-img-top"
        height="100%"
        src="https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg"
        width="auto"
      />
    </div>
    <div
      class="card-body"
    >
      <h5
        class="card-title"
      >
        Samsung XG-900 55 inch LCD TV
      </h5>
      <h6
        class="mb-2 text-muted card-subtitle"
      >
        <em>
          Latest generation smart televison 2021
        </em>
      </h6>
      <p
        class="card-text"
      >
        <span
          class="sc-bdfBwQ ehKgzr"
        >
          <span
            class="sc-gsTCUz cxQGHo price"
          >
            $400.40
          </span>
        </span>
      </p>
    </div>
  </div>
</DocumentFragment>
`;
